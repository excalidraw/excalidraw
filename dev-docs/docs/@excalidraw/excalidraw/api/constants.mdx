# Constants

### FONT_FAMILY

**How to use**

```js
import { FONT_FAMILY } from "@excalidraw/excalidraw";
```

`FONT_FAMILY` contains all the font families used in `Excalidraw` as explained below

| Font Family | Description            |
| ----------- | ---------------------- |
| `Virgil`    | The `handwritten` font |
| `Helvetica` | The `Normal` Font      |
| `Cascadia`  | The `Code` Font        |

Defaults to `FONT_FAMILY.Virgil` unless passed in `initialData.appState.currentItemFontFamily`.

### THEME

**How to use**

```js
import { THEME } from "@excalidraw/excalidraw";
```

`THEME` contains all the themes supported by `Excalidraw` as explained below

| Theme   | Description       |
| ------- | ----------------- |
| `LIGHT` | The `light` theme |
| `DARK`  | The `Dark` theme  |

Defaults to `THEME.LIGHT` unless passed in `initialData.appState.theme`

### MIME_TYPES

[`MIME_TYPES`](https://github.com/excalidraw/excalidraw/blob/master/src/constants.ts#L101) contains all the mime types supported by `Excalidraw`.

**How to use **

```js
import { MIME_TYPES } from "@excalidraw/excalidraw";
```
