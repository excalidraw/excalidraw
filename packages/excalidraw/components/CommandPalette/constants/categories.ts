export const DEFAULT_CATEGORIES = {
  app: "App",
  export: "Export",
  tools: "Tools",
  editor: "Editor",
  elements: "Elements",
  links: "Links",
};

export const CATEGORIES_ORDER = {
  [DEFAULT_CATEGORIES.app]: 1,
  [DEFAULT_CATEGORIES.export]: 2,
  [DEFAULT_CATEGORIES.editor]: 3,
  [DEFAULT_CATEGORIES.tools]: 4,
  [DEFAULT_CATEGORIES.elements]: 5,
  [DEFAULT_CATEGORIES.links]: 6,
};
