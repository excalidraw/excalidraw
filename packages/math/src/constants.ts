export const PRECISION = 10e-5;

// Legendre-Gauss abscissae (x values) and weights for n=24
// Refeerence: https://pomax.github.io/bezierinfo/legendre-gauss.html
export const LegendreGaussN24TValues = [
  -0.0640568928626056260850430826247450385909,
  0.0640568928626056260850430826247450385909,
  -0.1911188674736163091586398207570696318404,
  0.1911188674736163091586398207570696318404,
  -0.3150426796961633743867932913198102407864,
  0.3150426796961633743867932913198102407864,
  -0.4337935076260451384870842319133497124524,
  0.4337935076260451384870842319133497124524,
  -0.5454214713888395356583756172183723700107,
  0.5454214713888395356583756172183723700107,
  -0.6480936519369755692524957869107476266696,
  0.6480936519369755692524957869107476266696,
  -0.7401241915785543642438281030999784255232,
  0.7401241915785543642438281030999784255232,
  -0.8200019859739029219539498726697452080761,
  0.8200019859739029219539498726697452080761,
  -0.8864155270044010342131543419821967550873,
  0.8864155270044010342131543419821967550873,
  -0.9382745520027327585236490017087214496548,
  0.9382745520027327585236490017087214496548,
  -0.9747285559713094981983919930081690617411,
  0.9747285559713094981983919930081690617411,
  -0.9951872199970213601799974097007368118745,
  0.9951872199970213601799974097007368118745,
];

export const LegendreGaussN24CValues = [
  0.1279381953467521569740561652246953718517,
  0.1279381953467521569740561652246953718517,
  0.1258374563468282961213753825111836887264,
  0.1258374563468282961213753825111836887264,
  0.121670472927803391204463153476262425607,
  0.121670472927803391204463153476262425607,
  0.1155056680537256013533444839067835598622,
  0.1155056680537256013533444839067835598622,
  0.1074442701159656347825773424466062227946,
  0.1074442701159656347825773424466062227946,
  0.0976186521041138882698806644642471544279,
  0.0976186521041138882698806644642471544279,
  0.086190161531953275917185202983742667185,
  0.086190161531953275917185202983742667185,
  0.0733464814110803057340336152531165181193,
  0.0733464814110803057340336152531165181193,
  0.0592985849154367807463677585001085845412,
  0.0592985849154367807463677585001085845412,
  0.0442774388174198061686027482113382288593,
  0.0442774388174198061686027482113382288593,
  0.0285313886289336631813078159518782864491,
  0.0285313886289336631813078159518782864491,
  0.0123412297999871995468056670700372915759,
  0.0123412297999871995468056670700372915759,
];
