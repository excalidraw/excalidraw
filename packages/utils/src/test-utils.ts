import { diffStringsUnified } from "jest-diff";

expect.extend({
  toCloselyEqualPoints(received, expected, precision) {
    if (!Array.isArray(received) || !Array.isArray(expected)) {
      throw new Error("expected and received are not point arrays");
    }

    const COMPARE = 1 / Math.pow(10, precision || 2);
    const pass = expected.every(
      (point, idx) =>
        Math.abs(received[idx]?.[0] - point[0]) < COMPARE &&
        Math.abs(received[idx]?.[1] - point[1]) < COMPARE,
    );

    if (!pass) {
      return {
        message: () => ` The provided array of points are not close enough.

${diffStringsUnified(
  JSON.stringify(expected, undefined, 2),
  JSON.stringify(received, undefined, 2),
)}`,
        pass: false,
      };
    }

    return {
      message: () => `expected ${received} to not be close to ${expected}`,
      pass: true,
    };
  },
});
