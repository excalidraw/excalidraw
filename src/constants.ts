import { FontFamily } from "./element/types";

export const APP_NAME = "Excalidraw";

export const DRAGGING_THRESHOLD = 10; // 10px
export const LINE_CONFIRM_THRESHOLD = 10; // 10px
export const ELEMENT_SHIFT_TRANSLATE_AMOUNT = 5;
export const ELEMENT_TRANSLATE_AMOUNT = 1;
export const TEXT_TO_CENTER_SNAP_THRESHOLD = 30;
export const SHIFT_LOCKING_ANGLE = Math.PI / 12;
export const CURSOR_TYPE = {
  TEXT: "text",
  CROSSHAIR: "crosshair",
  GRABBING: "grabbing",
  POINTER: "pointer",
  MOVE: "move",
  AUTO: "",
};
export const POINTER_BUTTON = {
  MAIN: 0,
  WHEEL: 1,
  SECONDARY: 2,
  TOUCH: -1,
};

export enum EVENT {
  COPY = "copy",
  PASTE = "paste",
  CUT = "cut",
  KEYDOWN = "keydown",
  KEYUP = "keyup",
  MOUSE_MOVE = "mousemove",
  RESIZE = "resize",
  UNLOAD = "unload",
  BLUR = "blur",
  DRAG_OVER = "dragover",
  DROP = "drop",
  GESTURE_END = "gestureend",
  BEFORE_UNLOAD = "beforeunload",
  GESTURE_START = "gesturestart",
  GESTURE_CHANGE = "gesturechange",
  POINTER_MOVE = "pointermove",
  POINTER_UP = "pointerup",
  STATE_CHANGE = "statechange",
  WHEEL = "wheel",
  TOUCH_START = "touchstart",
  TOUCH_END = "touchend",
  HASHCHANGE = "hashchange",
  VISIBILITY_CHANGE = "visibilitychange",
}

export const ENV = {
  TEST: "test",
  DEVELOPMENT: "development",
};

export const CLASSES = {
  SHAPE_ACTIONS_MENU: "App-menu__left",
};

// 1-based in case we ever do `if(element.fontFamily)`
export const FONT_FAMILY = {
  1: "Virgil",
  2: "Helvetica",
  3: "Cascadia",
} as const;

export const WINDOWS_EMOJI_FALLBACK_FONT = "Segoe UI Emoji";

export const DEFAULT_FONT_SIZE = 20;
export const DEFAULT_FONT_FAMILY: FontFamily = 1;
export const DEFAULT_TEXT_ALIGN = "left";
export const DEFAULT_VERTICAL_ALIGN = "top";
export const DEFAULT_VERSION = "{version}";

export const CANVAS_ONLY_ACTIONS = ["selectAll"];

export const GRID_SIZE = 20; // TODO make it configurable?

export const MIME_TYPES = {
  excalidraw: "application/vnd.excalidraw+json",
  excalidrawlib: "application/vnd.excalidrawlib+json",
};

export const STORAGE_KEYS = {
  LOCAL_STORAGE_LIBRARY: "excalidraw-library",
};

// time in milliseconds
export const TAP_TWICE_TIMEOUT = 300;
export const TOUCH_CTX_MENU_TIMEOUT = 500;
export const TITLE_TIMEOUT = 10000;
export const TOAST_TIMEOUT = 5000;
export const VERSION_TIMEOUT = 30000;

export const ZOOM_STEP = 0.1;

// Report a user inactive after IDLE_THRESHOLD milliseconds
export const IDLE_THRESHOLD = 60_000;
// Report a user active each ACTIVE_THRESHOLD milliseconds
export const ACTIVE_THRESHOLD = 3_000;
