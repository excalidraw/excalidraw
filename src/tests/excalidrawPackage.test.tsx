import React from "react";
import { fireEvent, GlobalTestState, render } from "./test-utils";
import Excalidraw from "../packages/excalidraw/index";
import { queryByText, queryByTestId } from "@testing-library/react";
import { GRID_SIZE } from "../constants";

const { h } = window;

describe("<Excalidraw/>", () => {
  describe("Test zenModeEnabled prop", () => {
    it('should show exit zen mode button when zen mode is set and zen mode option in context menu when zenModeEnabled is "undefined"', async () => {
      const { container } = await render(<Excalidraw />);
      expect(
        container.getElementsByClassName("disable-zen-mode--visible").length,
      ).toBe(0);
      expect(h.state.zenModeEnabled).toBe(false);

      fireEvent.contextMenu(GlobalTestState.canvas, {
        button: 2,
        clientX: 1,
        clientY: 1,
      });
      const contextMenu = document.querySelector(".context-menu");
      fireEvent.click(queryByText(contextMenu as HTMLElement, "Zen mode")!);
      expect(h.state.zenModeEnabled).toBe(true);
      expect(
        container.getElementsByClassName("disable-zen-mode--visible").length,
      ).toBe(1);
    });

    it("should not show exit zen mode button and zen mode option in context menu when zenModeEnabled is set", async () => {
      const { container } = await render(<Excalidraw zenModeEnabled={true} />);
      expect(
        container.getElementsByClassName("disable-zen-mode--visible").length,
      ).toBe(0);
      expect(h.state.zenModeEnabled).toBe(true);

      fireEvent.contextMenu(GlobalTestState.canvas, {
        button: 2,
        clientX: 1,
        clientY: 1,
      });
      const contextMenu = document.querySelector(".context-menu");
      expect(queryByText(contextMenu as HTMLElement, "Zen mode")).toBe(null);
      expect(h.state.zenModeEnabled).toBe(true);
      expect(
        container.getElementsByClassName("disable-zen-mode--visible").length,
      ).toBe(0);
    });
  });

  describe("Test gridModeEnabled prop", () => {
    it('should show grid mode in context menu when gridModeEnabled is "undefined"', async () => {
      const { container } = await render(<Excalidraw />);
      expect(h.state.gridSize).toBe(null);

      expect(
        container.getElementsByClassName("disable-zen-mode--visible").length,
      ).toBe(0);
      fireEvent.contextMenu(GlobalTestState.canvas, {
        button: 2,
        clientX: 1,
        clientY: 1,
      });
      const contextMenu = document.querySelector(".context-menu");
      fireEvent.click(queryByText(contextMenu as HTMLElement, "Show grid")!);
      expect(h.state.gridSize).toBe(GRID_SIZE);
    });

    it('should not show grid mode in context menu when gridModeEnabled is not "undefined"', async () => {
      const { container } = await render(
        <Excalidraw gridModeEnabled={false} />,
      );
      expect(h.state.gridSize).toBe(null);

      expect(
        container.getElementsByClassName("disable-zen-mode--visible").length,
      ).toBe(0);
      fireEvent.contextMenu(GlobalTestState.canvas, {
        button: 2,
        clientX: 1,
        clientY: 1,
      });
      const contextMenu = document.querySelector(".context-menu");
      expect(queryByText(contextMenu as HTMLElement, "Show grid")).toBe(null);
      expect(h.state.gridSize).toBe(null);
    });
  });

  describe("Test theme prop", () => {
    it('should show the dark mode toggle when the theme prop is "undefined"', async () => {
      const { container } = await render(<Excalidraw />);
      expect(h.state.theme).toBe("light");

      const darkModeToggle = queryByTestId(container, "toggle-dark-mode");

      expect(darkModeToggle).toBeTruthy();
    });

    it('should not show the dark mode toggle when the theme prop is not "undefined"', async () => {
      const { container } = await render(<Excalidraw theme="dark" />);
      expect(h.state.theme).toBe("dark");

      expect(queryByTestId(container, "toggle-dark-mode")).toBe(null);
    });
  });
});
