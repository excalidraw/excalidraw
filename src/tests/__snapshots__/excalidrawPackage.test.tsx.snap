// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Excalidraw/> Test UIOptions prop Test canvasActions should not hide any UI element when canvasActions is "undefined" 1`] = `
<section
  aria-labelledby="canvasActions-title"
  class="zen-mode-transition"
>
  <h2
    class="visually-hidden"
    id="canvasActions-title"
  >
    Canvas actions
  </h2>
  <div
    class="Island"
    style="--padding: 2; z-index: 1;"
  >
    <div
      class="Stack Stack_vertical"
      style="--gap: 4;"
    >
      <div
        class="Stack Stack_horizontal"
        style="--gap: 1; justify-content: space-between;"
      >
        <button
          aria-label="Load"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="load-button"
          title="Load"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class="rtl-mirror"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
            >
              <path
                d="M572.694 292.093L500.27 416.248A63.997 63.997 0 0 1 444.989 448H45.025c-18.523 0-30.064-20.093-20.731-36.093l72.424-124.155A64 64 0 0 1 152 256h399.964c18.523 0 30.064 20.093 20.73 36.093zM152 224h328v-48c0-26.51-21.49-48-48-48H272l-64-64H48C21.49 64 0 85.49 0 112v278.046l69.077-118.418C86.214 242.25 117.989 224 152 224z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Save"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="save-button"
          title="Save"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class=""
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Save as"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--hide"
          data-testid="save-as-button"
          hidden=""
          title="Save as"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class=""
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M252 54L203 8a28 27 0 00-20-8H28C12 0 0 12 0 27v195c0 15 12 26 28 26h204c15 0 28-11 28-26V73a28 27 0 00-8-19zM130 213c-21 0-37-16-37-36 0-19 16-35 37-35 20 0 37 16 37 35 0 20-17 36-37 36zm56-169v56c0 4-4 6-7 6H44c-4 0-7-2-7-6V42c0-4 3-7 7-7h133l4 2 3 2a7 7 0 012 5z M296 201l87 95-188 205-78 9c-10 1-19-8-18-20l9-84zm141-14l-41-44a31 31 0 00-46 0l-38 41 87 95 38-42c13-14 13-36 0-50z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Export"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="export-button"
          title="Export"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class="rtl-mirror"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
            >
              <path
                d="M384 121.9c0-6.3-2.5-12.4-7-16.9L279.1 7c-4.5-4.5-10.6-7-17-7H256v128h128zM571 308l-95.7-96.4c-10.1-10.1-27.4-3-27.4 11.3V288h-64v64h64v65.2c0 14.3 17.3 21.4 27.4 11.3L571 332c6.6-6.6 6.6-17.4 0-24zm-379 28v-32c0-8.8 7.2-16 16-16h176V160H248c-13.2 0-24-10.8-24-24V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V352H208c-8.8 0-16-7.2-16-16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Reset the canvas"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="clear-canvas-button"
          title="Reset the canvas"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class=""
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        style="display: flex;"
      >
        <div
          style="position: relative;"
        >
          <div>
            <div
              class="color-picker-control-container"
            >
              <button
                aria-label="Canvas background"
                class="color-picker-label-swatch"
                style="--swatch-color: #ffffff;"
              />
              <label
                class="color-input-container"
              >
                <div
                  class="color-picker-hash"
                >
                  #
                </div>
                <input
                  aria-label="Canvas background"
                  class="color-picker-input"
                  spellcheck="false"
                  value="ffffff"
                />
              </label>
            </div>
          </div>
        </div>
        <div
          style="margin-inline-start: 0.25rem;"
        >
          <label
            class="ToolIcon ToolIcon_type_floating ToolIcon_size_M"
            data-testid="toggle-dark-mode"
            title="Dark mode"
          >
            <input
              aria-label="Dark mode"
              class="ToolIcon_type_checkbox ToolIcon_toggle_opaque"
              type="checkbox"
            />
            <div
              class="ToolIcon__icon"
            >
              <svg
                class="rtl-mirror"
                height="512"
                viewBox="0 0 512 512"
                width="512"
              >
                <path
                  d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </label>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`<Excalidraw/> Test UIOptions prop should not hide any UI element when the UIOptions prop is "undefined" 1`] = `
<section
  aria-labelledby="canvasActions-title"
  class="zen-mode-transition"
>
  <h2
    class="visually-hidden"
    id="canvasActions-title"
  >
    Canvas actions
  </h2>
  <div
    class="Island"
    style="--padding: 2; z-index: 1;"
  >
    <div
      class="Stack Stack_vertical"
      style="--gap: 4;"
    >
      <div
        class="Stack Stack_horizontal"
        style="--gap: 1; justify-content: space-between;"
      >
        <button
          aria-label="Load"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="load-button"
          title="Load"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class="rtl-mirror"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
            >
              <path
                d="M572.694 292.093L500.27 416.248A63.997 63.997 0 0 1 444.989 448H45.025c-18.523 0-30.064-20.093-20.731-36.093l72.424-124.155A64 64 0 0 1 152 256h399.964c18.523 0 30.064 20.093 20.73 36.093zM152 224h328v-48c0-26.51-21.49-48-48-48H272l-64-64H48C21.49 64 0 85.49 0 112v278.046l69.077-118.418C86.214 242.25 117.989 224 152 224z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Save"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="save-button"
          title="Save"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class=""
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Save as"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--hide"
          data-testid="save-as-button"
          hidden=""
          title="Save as"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class=""
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M252 54L203 8a28 27 0 00-20-8H28C12 0 0 12 0 27v195c0 15 12 26 28 26h204c15 0 28-11 28-26V73a28 27 0 00-8-19zM130 213c-21 0-37-16-37-36 0-19 16-35 37-35 20 0 37 16 37 35 0 20-17 36-37 36zm56-169v56c0 4-4 6-7 6H44c-4 0-7-2-7-6V42c0-4 3-7 7-7h133l4 2 3 2a7 7 0 012 5z M296 201l87 95-188 205-78 9c-10 1-19-8-18-20l9-84zm141-14l-41-44a31 31 0 00-46 0l-38 41 87 95 38-42c13-14 13-36 0-50z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Export"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="export-button"
          title="Export"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class="rtl-mirror"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
            >
              <path
                d="M384 121.9c0-6.3-2.5-12.4-7-16.9L279.1 7c-4.5-4.5-10.6-7-17-7H256v128h128zM571 308l-95.7-96.4c-10.1-10.1-27.4-3-27.4 11.3V288h-64v64h64v65.2c0 14.3 17.3 21.4 27.4 11.3L571 332c6.6-6.6 6.6-17.4 0-24zm-379 28v-32c0-8.8 7.2-16 16-16h176V160H248c-13.2 0-24-10.8-24-24V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V352H208c-8.8 0-16-7.2-16-16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
        <button
          aria-label="Reset the canvas"
          class="ToolIcon_type_button ToolIcon_size_m ToolIcon_type_button--show ToolIcon"
          data-testid="clear-canvas-button"
          title="Reset the canvas"
          type="button"
        >
          <div
            aria-hidden="true"
            class="ToolIcon__icon"
          >
            <svg
              aria-hidden="true"
              class=""
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
            >
              <path
                d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        style="display: flex;"
      >
        <div
          style="position: relative;"
        >
          <div>
            <div
              class="color-picker-control-container"
            >
              <button
                aria-label="Canvas background"
                class="color-picker-label-swatch"
                style="--swatch-color: #ffffff;"
              />
              <label
                class="color-input-container"
              >
                <div
                  class="color-picker-hash"
                >
                  #
                </div>
                <input
                  aria-label="Canvas background"
                  class="color-picker-input"
                  spellcheck="false"
                  value="ffffff"
                />
              </label>
            </div>
          </div>
        </div>
        <div
          style="margin-inline-start: 0.25rem;"
        >
          <label
            class="ToolIcon ToolIcon_type_floating ToolIcon_size_M"
            data-testid="toggle-dark-mode"
            title="Dark mode"
          >
            <input
              aria-label="Dark mode"
              class="ToolIcon_type_checkbox ToolIcon_toggle_opaque"
              type="checkbox"
            />
            <div
              class="ToolIcon__icon"
            >
              <svg
                class="rtl-mirror"
                height="512"
                viewBox="0 0 512 512"
                width="512"
              >
                <path
                  d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </label>
        </div>
      </div>
    </div>
  </div>
</section>
`;
